neps/nep-0488.md

```diff
@@ -0,0 +1,1537 @@
+- 0 points
+- points not on the curve
+- incorrectly encoded points
+- very long input
+
+***Tests References:***
+
+- Take the correct points on the curve from Ethereum tests[^47],[^48] and check the correctness after decompression
+- Randomly generate compressed points and check the equation correctness after decompression.
+
+***Error cases:***
+
+- The input length is not divided by 96
+- Too much memory is used
+- Field elements encoded incorrectly (see Curve points encoded section)
+- Any points not on the curve
+
+***Annotation:***
+
+```rust
+/// Decompress points on twisted BLS12-381 curve.
+///
+/// # Arguments
+///
+/// * `value` - sequence of pi points (x:Fp2) on twisted BLS-381 curve
+///    BLS12-381 is Y^2 = X^3 + (4 + i) curve over Fp2.
+///
+///   `value` is encoded as packed `[[u8; 96]]` slice.
+///
+/// # Output
+/// sequence of pi points (x: Fp2; y: Fp2) in decompress format: [([u8; 96], [u8; 96])]
+///
+/// # Errors
+///
+/// If `value_len + value_ptr` points outside the memory or the registers
+/// use more memory than the limit, the function returns `MemoryAccessViolation`.
+///
+/// If point coordinates are not on curve or `value.len()%96 != 0`,
+/// the function returns `BLS12381InvalidInput`.
+///
+/// # Cost
+///
+/// `base + write_register_base + write_register_byte * num_bytes +
+/// bls12381_decompress_g2_base + bls12381_decompress_g2_element * num_elements`
+pub fn bls12381_decompress_g2(&mut self, 
+                              value_len: u64, 
+                              value_ptr: u64, 
+                              register_id: u64) -> Result<()>;
+```
+
+## Reference Implementation
+
+First of all, for integration with nearcore, we are interested in libraries in the Rust language. The existing BLS12-381 implementations on Rust:
+
+1. ***Milagro Library*** [^29].
+2. ***BLST***  [^30][^31].
+3. ***Matter labs EIP-1962 implementation*** [^32]
+4. ***zCash origin implementation*** [^33]
+5. ***MCL Library*** [^34]
+6. ***FileCoin implementation*** [^35]
+7. ***zkCrypto*** [^36]
+
+To compile the list, we used the links from EIP-2537[^43], pairing-curves specification[^44], and an article with benchmarks[^45]. This list may be incomplete but should cover the core BLS12-381 implementations. In any case, to implement host functions from that NEP we will need to modify any of that libraries.
+
+In addition, there are implementations in other languages that are not so interesting to us in this context, but can be used as references:
+
+1. C++, ETH2.0 Client, ***Chia library***[^37]
+2. Haskell, ***Adjoint Lib***[^38]
+3. Go, ***Go-Ethereum***[^39]
+4. JavaScript, ***Noble JS***[^40]
+5. Go, ***Matter Labs Go EIP-1962 implementation***[^41]
+6. C++, ***Matter Labs Go EIP-1962 implementation***[^42]
+
+The draft implementation to nearcore you can find by this link[^54]. This implementation is based on *blst* library[^30]. This library one of the fastest[^45] and the most audited[^55].
```

Comment:
I think this doesn’t need to be in the NEP. I personally don’t think an audited assembly library is necessarily safe (as proved by openssl still regularly having vulnerabilities), but I haven’t checked alternatives and this discussion is for another time anyway, when thinking about landing the implementation :)

Reply:
@Ekleog-NEAR 

What exactly should not be in the NEP? I shouldn't give my opinion about which library is better to use? 

I thought, that the opposite, NEP should include the list of libraries and draft implementation. 

Yes, that is true, if the library is audited it doesn't mean, that it is correct. And I haven't done any analytics to choose the best library.

However, according to some evidence(such as it is quite well known, used in Ethereum, and has some audits), this library is not a bad choice.  And, another important thing: it is possible to implement these host-functions using this library. 

The message of this paragraph: the host functions from this NEP easy to implement by using existing not bad* libraries.    

We are not obliged to use this particular library and can choose something else if we have a reason for it

*not bad = works acceptably fast, the library is not dead, used by other big projects

Reply:
Right, I was just thinking about the "This library one of the fastest[^45] and the most audited[^55]." sentence; but it’s not a big deal anyway :) (I’m also not objecting to using this library for the production implementation)

Reply:
fixed: https://github.com/near/NEPs/pull/488/commits/1526529bc7559a75030f3b62a0d3bc73419f7b7c